homeassistant/components/plaato/sensor.py

```diff
@@ -31,134 +23,62 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     """Set up the Plaato sensor."""
 
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
+async def async_setup_entry(hass, entry, async_add_entities):
     """Set up Plaato from a config entry."""
     devices = {}
 
-    def get_device(device_id):
-        """Get a device."""
-        return hass.data[PLAATO_DOMAIN].get(device_id, False)
-
-    def get_device_sensors(device_id):
-        """Get device sensors."""
-        return hass.data[PLAATO_DOMAIN].get(device_id).get(PLAATO_DEVICE_SENSORS)
-
-    async def _update_sensor(device_id):
+    async def _update_sensor(device_id, sensor_data: PlaatoDevice):
         """Update/Create the sensors."""
-        if device_id not in devices and get_device(device_id):
-            entities = []
-            sensors = get_device_sensors(device_id)
-
-            for sensor_type in sensors:
-                entities.append(PlaatoSensor(device_id, sensor_type))
-
-            devices[device_id] = entities
-
-            async_add_entities(entities, True)
+        entry_data = hass.data[DOMAIN][entry.entry_id]
+        entry_data[SENSOR_DATA] = sensor_data
+
+        if entry.entry_id not in devices:
+            entry_data[DEVICE][DEVICE_ID] = device_id
+
+            entities = [
+                PlaatoSensor(hass.data[DOMAIN][entry.entry_id], sensor_type)
+                for sensor_type in sensor_data.sensors.keys()
+            ]
+            devices[entry.entry_id] = entities
```

Comment:
It would be a cleaner to store just the `unique_id` here since the entity itself isn't need for the signally.

Reply:
The same list is used when calling `async_add_entities` so I don't see how I could only store the `unique_id` here, or am I missing something?

Reply:
Since the same list is used in `async_add_entities` it would require to create the entity two times in a loop. Something like this

```python
entities = [
                PlaatoSensor(entry_data, sensor_type).unique_id
                for sensor_type in sensor_data.sensors.keys()
            ]
            devices[entry.entry_id] = entities
            async_add_entities(
                PlaatoSensor(entry_data, sensor_type)
                for sensor_type in sensor_data.sensors.keys()
            )
```

Or is there a smarter python way to do it?

Reply:
```python
async def async_setup_entry(hass, entry, async_add_entities):
    """Set up Plaato from a config entry."""
    entry_data = hass.data[DOMAIN][entry.entry_id]

    @callback
    def _async_update_from_webhook(device_id, sensor_data: PlaatoDevice):
        """Update/Create the sensors."""
        entry_data[SENSOR_DATA] = sensor_data

        if device_id not in entry_data[DEVICE][DEVICE_ID]:
            entry_data[DEVICE][DEVICE_ID] = device_id
            async_add_entities(
                [
                    PlaatoSensor(entry_data, sensor_type)
                    for sensor_type in sensor_data.sensors
                ]
            )
        else:
            for sensor_type in sensor_data.sensors:
                async_dispatcher_send(hass, f"{DOMAIN}_{device_id}_{sensor_type}")

    if entry.data[CONF_USE_WEBHOOK]:
        async_dispatcher_connect(hass, SENSOR_UPDATE, _async_update_from_webhook)
    else:
        coordinator = entry_data[COORDINATOR]
        async_add_entities(
            PlaatoSensor(entry_data, sensor_type, coordinator)
            for sensor_type in coordinator.data.sensors
        )

    return True
```

Reply:
- I think you can avoid storing devices since you know the format of the unique_id.  Bonus would be to make `unique_id` use a template in the property and here so you don't have two places constructing it
- Nothing was being awaited in the `_update_sensor` so it can be a callback instead.
- I think we can use `[]` instead of `get` for `CONF_USE_WEBHOOK` since it looks like it will always be set in the config flow.